### PR TITLE
docs: Add TDD documentation

### PR DESCRIPTION
**Pull request by mcjraquel**
This PR adds documentation for our Test-Driven Development (TDD) practices at BizKit.

### CODE REVIEW

**Review by Joshua-Hadap — APPROVED** [review]

**Review by jedreyes — APPROVED** [review]

**Review by domburza — APPROVED** [review]

File: content/docs/projects/tdd.md [hunk 1]
```diff
@@ -0,0 +1,602 @@
...
+title: "Test-Driven Development Guidelines"
+description: "Know more about our guidelines when writing and running unit tests."
+lead: "Author/s: Gab Barbudo, Celyn Raquel"
+date: 2023-07-05T14:30:18+08:00
+lastmod: 2023-07-05T14:30:18+08:00
+draft: false
+images: []
+menu:
+  docs:
+    parent: "projects"
```
Comment (kevingdc) [inline content/docs/projects/tdd.md]: Hmm, would it be better to put this under the same folder/section as the other standards?

File: content/docs/projects/tdd.md [hunk 2]
```diff
@@ -0,0 +1,602 @@
...
+# 1 Background
+
+**Test-driven development (TDD)** is a software development approach in which tests are written for a piece of code before the code itself is written. The idea is that by writing the tests first, you can define the desired behavior of the code and ensure that it works as expected.
+
+# 2 Key Features
+
+There are several key features of test-driven development that are important for beginners to understand:
+
+1. The idea that tests are written before the code itself is written. This ensures that the code is written to satisfy the requirements of the tests, rather than the other way around.
+2. The use of automated testing tools to run the tests and check the results. This allows for efficient and repeatable testing, and ensures that the tests are always run in the same way.
```
Comment (kevingdc) [inline content/docs/projects/tdd.md]:
```suggestion
2. The use of automated testing tools to run the tests and check the results. This allows for efficient and repeatable testing, and ensures that the tests always run in the same way.
```

File: content/docs/projects/tdd.md [hunk 3]
```diff
@@ -0,0 +1,602 @@
...
+
+3. <span style="color: blue;">Refactor</span>: Once the `add` function passes the test, you can refactor the code to make it more modular and readable. This might involve restructuring the code, adding comments, or making other improvements.
+
+```python
+def add(a, b):
+  # Return the sum of a and b
+  return a + b
+```
+
+In this example, the red, green, refactor approach is used to write a simple function called `add` that can be used to add two numbers together and return the result. The test is written first, and then the code is written to make the test pass. Finally, the code is refactored to improve its readability and maintainability.
```
Comment (kevingdc) [inline content/docs/projects/tdd.md]: Might also be good to note here that this is an iterative approach, where after refactoring, you would normally add more tests or update your tests to handle more scenarios. 

In the provided example, you might want to add more scenarios to the add test, like maybe looping through a list of different number cominations to add.

File: content/docs/projects/tdd.md [hunk 4]
```diff
@@ -0,0 +1,602 @@
...
+### 5.2.2 Basic Rules
+
+There are several basic rules that you should follow when writing unit tests in Frappe:
+
+1. Each test case should be a subclass of `unittest.TestCase` and should contain one or more test methods that begin with the word `test`.
+2. Test methods should use the various assertions provided by the `unittest` module to verify that the code under test is working correctly.
+3. Test cases should be isolated from one another. This means that each test should run independently and not depend on any other test to be run first.
+4. Test cases should be repeatable. This means that the same test should produce the same results every time it is run, regardless of the environment or other factors.
+5. Test cases should be thorough. This means that they should cover all relevant scenarios and test for all possible outcomes.
+6. Test cases should be easy to read and understand. This means that they should be well-structured and use clear, descriptive names for variables and methods.
```
Comment (kevingdc) [inline content/docs/projects/tdd.md]: Additional reminder: test code is still code. So it should follow the same standards as production code (i.e following DRY and KISS principles).

File: content/docs/projects/tdd.md [hunk 5]
```diff
@@ -0,0 +1,602 @@
...
+2. Avoid deleting documents and transactions to avoid errors when deleting documents linked to other documents.
+3. Start with a small, well-defined task. This will help you focus on the task at hand and avoid getting overwhelmed.
+4. Write a test for the task before you write the code. This will help you understand the requirements and constraints of the task, and it will give you a clear goal to work towards.
+5. Write the minimum amount of code needed to make the test pass. This will help you avoid over-engineering and keep your code simple and focused.
+6. Refactor your code to improve its design and readability. This will help you make your code more maintainable and easier to understand.
+7. Repeat the process for each new feature or task you want to implement. This will help you build a suite of tests that will give you confidence in the correctness and completeness of your code.
+8. Be persistent and keep practicing. Test-driven development can be challenging at first, but it becomes easier and more natural with time and practice.
```
Comment (kevingdc) [inline content/docs/projects/tdd.md]: Might also be good to add the tip to create a separate site for running the tests and sharing the instructions on how to do that.

File: content/docs/projects/tdd.md [hunk 6]
```diff
@@ -0,0 +1,602 @@
...
+
+# 6 Running Unit Tests
+
+This is the _exciting_ part! (lol is it really)
```
Comment (kevingdc) [inline content/docs/projects/tdd.md]: Love the humor

File: content/docs/projects/tdd.md [hunk 7]
```diff
@@ -0,0 +1,602 @@
...
+
+Follow these steps when running unit tests in your local instance:
+
+1. Run your unit tests individually first (using the `--test` flag).
+2. Once all your unit tests have pass individually, run the whole module/DocType if possible (using the `--doctype` or `--module` flag).
```
Comment (kevingdc) [inline content/docs/projects/tdd.md]:
```suggestion
2. Once all your unit tests have passed individually, run the whole module/DocType if possible (using the `--doctype` or `--module` flag).
```

File: content/docs/projects/tdd.md [hunk 8]
```diff
@@ -0,0 +1,602 @@
...
+3. If all tests run pass as well, run all BizKit-updated tests (using the `bench run-bizkit-tests` command). Make sure that your new tests are accounted for in the `updated_doctypes.txt` and/or `updated_modules.txt` file.
```
Comment (kevingdc) [inline content/docs/projects/tdd.md]:
```suggestion
3. If all tests pass as well, run all BizKit-updated tests (using the `bench run-bizkit-tests` command). Make sure that your new tests are accounted for in the `updated_doctypes.txt` and/or `updated_modules.txt` file.
```

File: content/docs/projects/tdd.md [hunk 9]
```diff
@@ -0,0 +1,602 @@
...
+4. If you encounter no error while running all BizKit-updated tests, you're **#MarkedAsSafe**!
```
Comment (kevingdc) [inline content/docs/projects/tdd.md]:
```suggestion
4. If you encounter no errors while running all BizKit-updated tests, you're **#MarkedAsSafe**!
```

**Review by kevingdc — CHANGES_REQUESTED** [review]

File: content/docs/projects/tdd.md (hunk 1 above)
Comment (mcjraquel) [reply content/docs/projects/tdd.md]: 856cf12959eea44b2307f78bbf6ce9b83a24a51b

File: content/docs/projects/tdd.md (hunk 2 above)
Comment (mcjraquel) [reply content/docs/projects/tdd.md]: 856cf12959eea44b2307f78bbf6ce9b83a24a51b

File: content/docs/projects/tdd.md (hunk 3 above)
Comment (mcjraquel) [reply content/docs/projects/tdd.md]: 856cf12959eea44b2307f78bbf6ce9b83a24a51b

File: content/docs/projects/tdd.md (hunk 4 above)
Comment (mcjraquel) [reply content/docs/projects/tdd.md]: 856cf12959eea44b2307f78bbf6ce9b83a24a51b

File: content/docs/projects/tdd.md (hunk 5 above)
Comment (mcjraquel) [reply content/docs/projects/tdd.md]: 856cf12959eea44b2307f78bbf6ce9b83a24a51b

File: content/docs/projects/tdd.md (hunk 7 above)
Comment (mcjraquel) [reply content/docs/projects/tdd.md]: 856cf12959eea44b2307f78bbf6ce9b83a24a51b

File: content/docs/projects/tdd.md (hunk 8 above)
Comment (mcjraquel) [reply content/docs/projects/tdd.md]: 856cf12959eea44b2307f78bbf6ce9b83a24a51b

File: content/docs/projects/tdd.md (hunk 9 above)
Comment (mcjraquel) [reply content/docs/projects/tdd.md]: 856cf12959eea44b2307f78bbf6ce9b83a24a51b

**Review by mcbsalceda — APPROVED** [review]

**Review by kevingdc — APPROVED** [review]

**Review by rover96 — APPROVED** [review]

**Pull request merged**